Explain deployment options for kyma functions (#13813)
kyma-project/kyma: docs/01-overview/main-areas/serverless/svls-01-function-deployment-options.md

---
title: Best Practices for Function Development
---

# Overview - Its all about Custom Resources

> Kyma serverless introduces a [`Function`](https://kyma-project.io/docs/kyma/latest/05-technical-reference/00-custom-resources/svls-01-function/) Custom Resource Definition (CRD) as an extension to the kubernetes API server.

Reviewer: ```suggestion
Kyma serverless introduces a [`Function`](../../05-technical-reference/00-custom-resources/svls-01-function/) Custom Resource Definition (CRD) as an extension to the kubernetes API server.
```

>>NOTE: Use `kyma init function --vscode` to generate json schema which can be used in VScode for autocompletion.
>>

Kyma CLI helps you run your code locally with a single `kyma run function` command. It runs the function using your local docker deamon with the same runtime docker context as if it was run in the kyma runtime. 
>>NOTE: Use `kyma run function` with `--hot-deploy` and spare yourself unnecessary restarts of the functions whenever you test a changed function logic. Also, use [`--debug` option](https://kyma-project.io/docs/kyma/latest/03-tutorials/00-serverless/svls-05-debug-function) to allow connecting with your favourite debugger.
>>

![kyma-cli-functions](../assets/kyma-cli-functions.png)

Having written and tested your function locally, simply deploy it to the kyma runtime with `kyma apply function` command used in the folder of your function's workspace. It will read the files, translate it to kubernetes manifests and deploy the function.

# Deploy via CI/CD

Kyma UI helps you getting started. Kyma CLI helps you iterate and develop functions. 

> But at the end of the day you probably want an automated deployment of your functions. Or rather, of your whole application, where functions are just part of it.

Reviewer: ```suggestion
But at the end of the day you may probably want an automated deployment of your Functions. Or rather, of your whole application, where Functions are just part of it.
```

> It all comes down to being able to deploy k8s applications on different kyma runtimes in a GitOps fashion and, for the sake of simplicity, the deployment approach for functions should not differ from deployment of other kubernetes workloads, config-maps or secrets.

Reviewer: ```suggestion
It all comes down to deployment of Kubernetes applications on different Kyma runtimes in a GitOps fashion and, for the sake of simplicity, the deployment approach for Functions should not differ from deployment of other Kubernetes workloads, Config Maps or Secrets.
```

> So in the end what you need is those yaml manifests for everything - including functions.

Reviewer: ```suggestion
So in the end what you need is those yaml manifests for everything - including Functions.
```

> Fortunatelly, Kyma CLI helps you generate the yaml mainfests matching your `config.yaml` file you crafted in earlier phase.

Reviewer: ```suggestion
Fortunately, Kyma CLI helps you generate the yaml mainfests matching your `config.yaml` file crafted before.
```

> Use `--dry-run` option of the `kyma apply function` command to generate kubernetes manifests that will include the function CR itself but also all the related CRs (i.e ApiRules, Subscriptions, etc).

Reviewer: ```suggestion
Use `--dry-run` option of the `kyma apply function` command to generate Kubernetes manifests that will include the Function CR itself but also all the related CRs (i.e. Api Rules, Subscriptions, etc).
```

```bash
kyma apply function --dry-run --ci -o yaml > my-function.yaml
```  

The generated manifest should be part of all the manifests that define your application and pushed to the git repository.

> Deploy everything in a consistent way either via CI/CD or via GitOps operators ( i.e [fluxcd](https://kyma-project.io/docs/kyma/latest/03-tutorials/00-serverless/svls-06-sync-function-with-gitops/) ) installed in your kyma runtime.

Reviewer: ```suggestion
Deploy everything in a consistent way either via CI/CD or via GitOps operators (e.g. [fluxcd](../../03-tutorials/00-serverless/svls-06-sync-function-with-gitops.md) installed on your Kyma runtime.
```

> >>NOTE: Source Function Code  directly from Git Repository

Reviewer: ```suggestion
>>NOTE: Source the Function code directly from git repository
```

> Kyma Functions come in two types : `git` and `inline`.

Reviewer: ```suggestion
Kyma Functions come in two types: `git` and `inline`.
```

> [Git type](https://kyma-project.io/docs/kyma/latest/03-tutorials/00-serverless/svls-02-create-git-function/) allows you to configure a git repository as a source of your function code instead of creating it "inline".

Reviewer: ```suggestion
[Git type](../../03-tutorials/00-serverless/svls-02-create-git-function.md) allows you to configure a git repository as a source of your Function code instead of creating it `inline`.
```

> This allows you to skip rendering of k8s manifests and deploying them everytime you made a change in the function code or dependencies. Simply push the changes to the referenced got repository and the serverless controller will rebuild the functions that is deployed in your kyma runtime. 

Reviewer: ```suggestion
This allows you to skip rendering of the Kubernetes manifests and deploying them each time you made a change in the Function code or dependencies. Simply push the changes to the referenced git repository and the Serverless controller will rebuild the Function that is deployed in your Kyma runtime. 
```

> Please have a look at this [example](https://github.com/kyma-project/examples/tree/main/incluster_eventing) that illustrates how you could setup your git project. Mind  the `k8s resources` folder with yaml manifests to be pushed to the kubernetes API server ( i.e via kubeclt in our  CI/CD or GitOps ) and `src` folder containing functions source code that are pulled direclty by kyma serverless to build new function images whenever the source content changes in the git repository.  

Reviewer: ```suggestion
Please have a look at this [example](https://github.com/kyma-project/examples/tree/main/incluster_eventing) that illustrates how you can setup your git project. Mind  the `k8s resources` folder with the yaml manifests to be pushed to the Kubernetes API server ( e.g. via kubectl in our CI/CD or GitOps ) and `src` folder containing the Functions' source code. They are pulled directly by Kyma Serverless to build new Function images whenever the source content changes in the git repository.  
```
